### DISCUSSION TITLE
How to properly use setViewport and setScissorRect?

### DISCUSSION THREAD

**Discussion by smolck** [gpuweb/gpuweb#1506] [Q&A] [answered]
Hello! So, I'm trying to port the rendering of a text editor I've been working on to WebGPU (see https://github.com/smolck/uivonim/pull/165 if you're interested), and I'm running into issues with the details of the port (most if not all of which are probably due to my lack of graphics programming knowledge 😅 ). 

Specifically, one of the things I'm having trouble understanding/finding information on is how to use `setViewport` and `setScissorRect`. I keep getting errors about the viewport or scissor rect needing "to be contained in the render targets" but I can't seem to find any information on how to achieve that, or what that error even means. My assumption is that it means that the swapchain texture to which I'm rendering has smaller dimensions than the viewport/scissor rect I'm trying to set, but I haven't been able to confirm or deny this, or find any examples on updating the dimensions of the swapchain texture to fix this if that assumption is correct.

From reading https://github.com/austinEng/webgpu-samples/issues/79#issuecomment-777772685, maybe I just need to re-call `configureSwapChain` when the canvas is resized? Or what am I missing? Also, are there any examples somewhere which demonstrate how to properly use these functions that I happened to miss? I couldn't find any usage of `setViewport` or `setScissorRect` in [austinEng/webgpu-samples](https://github.com/austinEng/webgpu-samples), specifically.

Thanks in advance!

**Comment by kainino0x** [comment, marked as answer]
FYI, here's the error logic from Dawn (Chromium):
https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/dawn_native/RenderPassEncoder.cpp;l=151;drc=05045e0ad8588d7c679ec93d12b9921104f7bac8
Unfortunately I don't think we currently have a way to inspect what the actual size of the swap chain texture is, but indeed this means your viewport/scissor rect are larger than the swap chain texture.

> From reading [austinEng/webgpu-samples#79 (comment)](https://github.com/austinEng/webgpu-samples/issues/79#issuecomment-777772685), maybe I just need to re-call `configureSwapChain` when the canvas is resized?

~~I'm not sure whether Chromium requires that or not (@austinEng?).~~ Unfortunately, I don't think the spec has yet made a call on how that should work. I am not sure yet what the interactions should be between configureSwapChain, canvas resizes, and present.

**Reply by austinEng** [reply]
As Kai said, I'm not sure what the spec is supposed to be. But right now in Chromium, I don't think you need to call `configureSwapChain` again. Perhaps my previous comment was wrong.

Here's a little demo https://jsfiddle.net/emf5cn0k/
Passing in `canvas.width` and `canvas.height` to both `setScissorRect` and `setViewport` worsk. If you say `canvas.width + 1` you'll get the validation errors you're talking about.

**Reply by Kangz** [reply]
@kainino0x I'm pretty sure the spec will say that the texture you get has its size set with the canvas' width and height at the moment you created it.

**Reply by kvark** [reply]
Yes, I don't think there is any meaningful alternative to this.

**Reply by kainino0x** [reply]
The alternative I had in mind is that if you set canvas width and height, it would invalidate any existing swap chain texture and you would have to get a new one.

The semantic you described seems better and I'm pretty sure it works fine; you just can have one frame where the target size doesn't match the canvas and it gets composited in a weird way (IIRC this can happen for WebGL as well, I don't remember how)
